### PR TITLE
GEOS-7844: move Importer to use high level api

### PR DESCRIPTION
**Pull request by DBlasby**
This PR does two things;
a) splits the (quite long) loadIntoDataStore() method into two parts - setup and transferring features
b) provide two implementation for moving features one based on the low-level api (read-a-feature write-a-feature) and one based on the high-level api (FeatureCollection/featureStore.addFeatures).

For the non-datastore formats (GeoJSON, GML, KML), this PR will not affect functionality.  However, for Datastore-backed data sources, this uses the FeatureCollection/featureStore.addFeatures() method.

The original feature transformation (cf copyFeaturesLowLevelAPI) has been re-implemented inside a FeatureCollection/FeatureIterator.

The FeatureIterator provides two pieces of functionality;
a) cancelability (cf CancelableFeatureIterator)
b) feature xformation (cf TransformingFeatureIterator)

The actual functionality remains the same - this just does the same thing inside the FeatureIterator instead of in a read-a-feature write-a-feature while loop.

Notes;
a) I re-formated the modified code
b) there was already existing adequate test coverage

### CODE REVIEW

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 1]
```diff
@@ -1172,192 +1175,107 @@ boolean doPostTransform(ImportTask task, ImportData data, TransformChain tx) {
...
-        FeatureWriter writer = null;
+
         // using this exception to throw at the end
         Exception error = null;
         Transaction transaction = new DefaultTransaction();
         try {
-            reader = format.read(data, task);
 
-            SimpleFeatureType featureType = (SimpleFeatureType) reader.getFeatureType();
+            SimpleFeatureType featureType = (SimpleFeatureType) task.getMetadata()
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Is that really the api? That looks pretty informal ...

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 2]
```diff
@@ -1172,192 +1175,107 @@ boolean doPostTransform(ImportTask task, ImportData data, TransformChain tx) {
...
+
                 if (!uniquifiedFeatureTypeName.equals(featureTypeName)) {
-                    //update the metadata
+                    // update the metadata
                     task.getLayer().getResource().setName(uniquifiedFeatureTypeName);
                     task.getLayer().getResource().setNativeName(uniquifiedFeatureTypeName);
-                    
-                    //retype
+
+                    // retype
```
Comment (tbarsballe) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: While I appreciate the formatting fixes, such changes should really be a separate commit. Mixing format and code changes makes a PR rather more difficult to review.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 3]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
         // finally, throw any error
         if (error != null) {
             throw error;
         }
     }
 
+    private Exception copyFeaturesHighLevelAPI(ImportData data, ImportTask task,
+            DataStoreFormat format, DataStore dataStore, Transaction transaction,
+            String featureTypeName, String uniquifiedFeatureTypeName,
+            FeatureDataConverter featureDataConverter, VectorTransformChain tx) throws IOException {
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Would prefer "copyFromFeatureSource" since we do not have a formal high / low / internal api.

Is it worth untangling some of these featureTypeName, dataStore, transaction parameters by passing in a single configured featureStore to this method?

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 4]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
+        Exception error = null;
+        ProgressMonitor monitor = task.progress();
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Should this be a sub progress monitor? Just asking I am not sure ...

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 5]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
+        try {
+            long startTime = System.currentTimeMillis();
+            task.clearMessages();
+
+            task.setTotalToProcess(format.getFeatureCount(task.getData(), task));
+            LOGGER.info("begining import - highlevel api");
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Set logging level to fine for this one

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 6]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
+
+            FeatureSource fs = format.getFeatureSource(data, task);
+            FeatureCollection fc = fs.getFeatures();
+
+            FeatureStore featureStore = (FeatureStore) dataStore
+                    .getFeatureSource(uniquifiedFeatureTypeName);
+            featureStore.setTransaction(transaction);
+
+            fc = new TransformingCancelableInteratingFC(fc, monitor, featureDataConverter,
+                    featureStore.getSchema(), tx, task, dataStore);
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Would like to follow class name conventions here; there should be some base classes to start from.
- fc this is your original featureCollection
- monitor - this could only be used once, not suitable for a reusable FeatureCollection implementation
- tx, datastore - this is already known to the fc why so many parameters?

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 7]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
+
+        if (error != null || monitor.isCanceled()) {
+            // all sub exceptions in this catch block should be logged, not thrown
+            // as the triggering exception will be thrown
+
+            // failure, rollback transaction
+            try {
+                transaction.rollback();
+            } catch (Exception e1) {
+                LOGGER.log(Level.WARNING, "Error rolling back transaction", e1);
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: provide some context here as this will be the only message end users will see in their logs.
```
LOGGER.log(Level.WARNING,"Unable to load data into "+ uniquifiedFeatureTypeName+", rolling back data insert:"+e1, e1);
```

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 8]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
+            }
+
+            // attempt to drop the type that was created as well
+            try {
+                dropSchema(dataStore, featureTypeName);
+            } catch (Exception e1) {
+                LOGGER.log(Level.WARNING, "Error dropping schema in rollback", e1);
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: interesting, createSchema does not allow the table to be crated on a transaction.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 9]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
+            }
+        }
+
+        return error;
+    }
+
+    Exception copyFeaturesLowLevelAPI(FeatureReader reader, ImportTask task, VectorFormat format,
+            DataStore dataStore, Transaction transaction, String featureTypeName,
+            String uniquifiedFeatureTypeName, FeatureDataConverter featureDataConverter,
+            VectorTransformChain tx) throws IOException {
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Same feedback as above:
- prefer "copyFromFeatureReader"
- that is a lot of parameters

aside: There is a DataUtilities method that can wrap a FeatureReader as a FeatureCollection, but it may store the results in memory. uDig has a single use feature collection wrapper which would be suitable but is a pretty brutal solution.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+import org.opengis.feature.type.FeatureType;
+
+import com.vividsolutions.jts.geom.Geometry;
+
+/**
+ * FeatureCollection that does two things required by the importer; a) provide cancel functionality b) Do some FeatureType Transforming
+ *
+ * This class is simply wraps the FeatureIterator with two iterators wrappers that provide the above functionality.
+ */
+public class TransformingCancelableInteratingFC extends DecoratingFeatureCollection {
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java]: Nice, you found DecoratingFeatureCollection. 

- You may wish to make this package visible, rather than commit to additional public api
- Can we rename this to "ImportTransformFeatureCollection" or something to follow the naming convention.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+    ProgressMonitor monitor;
+
+    FeatureDataConverter featureDataConverter;
+
+    FeatureType resultingFT;
+
+    VectorTransformChain tx;
+
+    ImportTask task;
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java]: This task can provide a monitor as needed, no need to pass it in?

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 4 above)
Comment (tbarsballe) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: No, I don't believe it should

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 7 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: This is the original implementation

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 3 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: I updated the name.

I don't think untangling is going to help at all - there's just going to be another method added that pretty much does what this one does.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 6 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Not sure what you're saying here...

File: src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java (hunk 10 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java]: I changed visibility and name.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java (hunk 11 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/TransformingCancelableInteratingFC.java]: fixed

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 7 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: updated

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 5 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: I updated this, as well as the other  one.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 6 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: If you look at the VectorTransformChain, it requires a lot of classes.
tx - is the VectorTransformChain (I'll rename - that was it's original name)
datastore - this is actually the destination datastore (I'll rename - that was it's original name)

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java [hunk 12]
```diff
@@ -1369,16 +1287,146 @@ else if (isPostGISDataStore(dataStore)) {
...
+            FeatureStore featureStore = (FeatureStore) dataStoreDestination
+                    .getFeatureSource(uniquifiedFeatureTypeName);
+            featureStore.setTransaction(transaction);
+
+            fc = new ImportTransformFeatureCollection(fc, featureDataConverter,
+                    featureStore.getSchema(), tx, task, dataStoreDestination);
+
+            featureStore.addFeatures(fc);
+
+        } catch (Exception e) {
```
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Think we want to catch RuntimeExceptions here as well - or use a try/finally block to ensure the transaction is either committed or rolled back.

**Comment by jodygarnett** [comment]
Excellent Dave, I had one thing I was unsure of with try/finally, but this is looking great.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 12 above)
Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Yes - further up the chain.  CF line 1333-1344.  It rolls back the transaction as well as dropping the created schema.

**Comment by tbarsballe** [comment]
I would still like to make sure we squash the commits before merging, but yes, this is looking good.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java (hunk 12 above)
Comment (jodygarnett) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: Sorry Dave I ment Throwable, we need to catch Throwable here or we have a chance of missing the clean up code on 1333-1344

Comment (DBlasby) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Importer.java]: changed

**Comment by aaime** [comment]
So, before we had a single code path to maintain using readers, and now there are two, one using readers, one using feature collections? Looks more code, more options to understand, where is the improvement?
